btprotocol/protocol_test.go

```diff
@@ -152,3 +152,96 @@ func TestUnmarshalPortMsg(t *testing.T) {
 		t.FailNow()
 	}
 }
+
+func TestMarshalBinaryPooled(t *testing.T) {
+	messages := []Message{
+		NewKeepAlive(),
+		NewHavePiece(42),
+		NewChoked(),
+		NewUnchoked(),
+		NewInterested(true),
+		NewInterested(false),
+		{Type: Request, Index: 5, Begin: 0, Length: 16384},
+		{Type: Cancel, Index: 5, Begin: 0, Length: 16384},
+		NewPiece(1, 0, []byte("test piece data")),
+		NewPort(6881),
+		NewHaveAll(),
+		NewHaveNone(),
+		NewAllowedFast(10),
+	}
+
+	for _, msg := range messages {
+		t.Run(msg.Type.String(), func(t *testing.T) {
+			expected, err := msg.MarshalBinary()
+			assert.NoError(t, err)
+
+			buf := AcquireBuffer()
+			defer ReleaseBuffer(buf)
+			actual, err := msg.MarshalBinaryPooled(buf)
+			assert.NoError(t, err)
+			assert.Equal(t, expected, actual)
+		})
+	}
+}
+
+func BenchmarkMarshalBinary(b *testing.B) {
+	b.Run("piece/16KB", func(b *testing.B) {
+		msg := NewPiece(1, 0, make([]byte, 16384))
+		b.ReportAllocs()
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			_, _ = msg.MarshalBinary()
```

Comment:
you can just update MArshalBinary and inline the pool in this test, and pass a shared buffer in the other test to compare relative performance.